weather oracle (#21)
PolkadotStudy/polkadot.study: tutorials/substrate-inherents-tutorial/2-client.mdx

# Client
## Inherent Data Provider
A block producing node relies on an ```InherentDataProvider``` whenever it is interested to pass

> data to Runtime upon authoring a new block.

Reviewer: ```suggestion
data to the runtime upon authoring a new block.
```

> An ```InherentDataProvider``` trait implementor utilizes ```provide_inherent_data``` method to put data onto 

Reviewer: ```suggestion
An ```InherentDataProvider``` trait implementor utilizes the ```provide_inherent_data``` method to put data into
```

> storage location under a key both the Client and the Runtime share.

Reviewer: ```suggestion
storage under a key both the Client and the Runtime share.
```

As the first step you'll define a struct implementing the aforementioned ```InherentDataProvider``` trait.

> Please head to ./node/src folder and add a file ```inherent_data_provider.rs``` with the following code:

Reviewer: ```suggestion
Please head to `./node/src` folder and add a file ```inherent_data_provider.rs``` with the following code:
```

		if *identifier == INHERENT_IDENTIFIER {
            Some(Err(sp_inherents::Error::Application(Box::from(std::format!("Error processing inherent: {:?}", error)))))
		} else {
            None
        }
	}
}

``` 

This code references a few data types that you'll later define on the Runtime side.
Let's have a look at them.

The line 
```rust
use node_template_runtime::{InherentDataType}
```
references the type of the inherent as it is understood by the Runtime. Your inherent data provider will hold
an instance of this type which it will pass to the Runtime in a SCALE-encoded format.

Then we have
```rust
use node_template_runtime::inherents_example::{INHERENT_IDENTIFIER}
```
which is the key under which the block producer passes inherent data to the Runtime. 

The struct ```ExternalDataInherentProvider``` holds the data you want the block producer to pass to the Runtime
and implements the trait ```InherentDataProvider``` by putting it onto storage from which the data will be then
fetched and decoded by the Runtime.

## Create Inherent Data Provider
Now when your inherent data provider is ready it must be used by the block production.

> Inherent data provider instances are created by a closure passed to consensus block production module.

Reviewer: ```suggestion
Inherent data provider instances are created by a closure passed to the consensus block production module.
```

Please head to the file ```node/src/service.rs``` and locate the 
```rust title=node/src/service.rs
create_inherent_data_providers: move |_, ()| async move {
    let timestamp = sp_timestamp::InherentDataProvider::from_system_time();

    let slot =
        sp_consensus_aura::inherents::InherentDataProvider::from_timestamp_and_slot_duration(
            *timestamp,
            slot_duration,
        );

    Ok((slot, timestamp))
},
```
code snippet.

There are two identical snippets in the file, at the moment we are going to concentrate
on that referred by ```sc_consensus_aura::start_aura```.

> We see that there are already two inherent data providers returned by the closure which

Reviewer: ```suggestion
You can see that there are already two inherent data providers returned by the closure which
```

> Aura uses to pass inherent data to the Runtime when it produces a block.

Reviewer: ```suggestion
AURA uses to pass inherent data to the Runtime when it produces a block.
```

> Aura passes the current time and slot inherents to the Runtime.

Reviewer: ```suggestion
AURA passes the current time and slot inherents to the Runtime.
```
